File: cases/approval_callback_bypass/approval.py
import os

from agents import Agent, ShellTool


SHELL_AUTO_APPROVE = os.environ.get("SHELL_AUTO_APPROVE") == "1"
UNUSED_AUTO_APPROVE = os.environ.get("UNUSED_AUTO_APPROVE") == "1"


async def prompt_shell_approval() -> bool:
    if SHELL_AUTO_APPROVE:
        return True
    return False


async def approval_wrapper(_context, _item) -> bool:
    return await prompt_shell_approval()


async def unused_prompt() -> bool:
    if UNUSED_AUTO_APPROVE:
        return True
    return False


async def safe_callback(_context, _item) -> bool:
    return False


operator = Agent(
    name="operator",
    tools=[
        ShellTool(
            executor=object(),
            needs_approval=True,
            on_approval=approval_wrapper,
        ),
        ShellTool(
            executor=object(),
            needs_approval=True,
            on_approval=safe_callback,
        ),
        ShellTool(
            environment={"type": "container_auto"},
            needs_approval=True,
            on_approval=approval_wrapper,
        ),
    ],
)

File: cases/approval_callback_bypass/approval.ts
import { Agent, applyPatchTool, shellTool } from '@openai/agents';

async function promptShellApproval() {
  if (process.env.SHELL_AUTO_APPROVE === '1') {
    return true;
  }
  return false;
}

async function promptPatchApproval() {
  if (process.env.APPLY_PATCH_AUTO_APPROVE === '1') {
    return true;
  }
  return false;
}

async function unusedPrompt() {
  if (process.env.UNUSED_AUTO_APPROVE === '1') {
    return true;
  }
  return false;
}

const operator = new Agent({
  name: 'operator',
  tools: [
    shellTool({
      shell: {},
      needsApproval: true,
      onApproval: async () => ({ approve: await promptShellApproval() }),
    }),
    applyPatchTool({
      editor: {},
      needsApproval: true,
      onApproval: async () => ({ approve: await promptPatchApproval() }),
    }),
    shellTool({
      environment: { type: 'container_auto' },
      needsApproval: true,
      onApproval: async () => ({ approve: await promptShellApproval() }),
    }),
    shellTool({
      shell: {},
      needsApproval: true,
      onApproval: async () => {
        void unusedPrompt;
        return { approve: false };
      },
    }),
  ],
});
